test/Dialect/Moore/sroa.mlir

```diff
@@ -1,58 +1,56 @@
-  // CHECK: %z = moore.variable : <i32>
-  %x = moore.variable : <i32>
-  %y = moore.variable : <i32>
-  %z = moore.variable : <i32>
-  moore.procedure always_comb {
-    // CHECK: %a = moore.variable : <i32>
-    // CHECK: %b = moore.variable : <i32>
-    // CHECK: %0 = moore.constant 1 : i32
-    // CHECK: %1 = moore.conversion %0 : !moore.i32 -> !moore.i32
-    // CHECK: moore.blocking_assign %a, %1 : i32
-    // CHECK: %2 = moore.constant 4 : i32
-    // CHECK: %3 = moore.conversion %2 : !moore.i32 -> !moore.i32
-    // CHECK: moore.blocking_assign %b, %3 : i32
-    // CHECK: %4 = moore.read %x
-    // CHECK: %5 = moore.constant 1 : i32
-    // CHECK: %6 = moore.add %4, %5 : i32
-    // CHECK: moore.blocking_assign %a, %6 : i32
-    // CHECK: %7 = moore.read %a
-    // CHECK: moore.blocking_assign %y, %7 : i32
-    // CHECK: %8 = moore.read %a
-    // CHECK: %9 = moore.constant 1 : i32
-    // CHECK: %10 = moore.add %8, %9 : i32
-    // CHECK: moore.blocking_assign %a, %10 : i32
-    // CHECK: %11 = moore.read %a
-    // CHECK: moore.blocking_assign %z, %11 : i32
-    %ii = moore.variable : <struct<{a: i32, b: i32}>>
-    %0 = moore.struct_extract_ref %ii, "a" : <struct<{a: i32, b: i32}>> -> <i32>
-    %1 = moore.constant 1 : i32
-    %2 = moore.conversion %1 : !moore.i32 -> !moore.i32
-    moore.blocking_assign %0, %2 : i32
-    %3 = moore.struct_extract_ref %ii, "b" : <struct<{a: i32, b: i32}>> -> <i32>
-    %4 = moore.constant 4 : i32
-    %5 = moore.conversion %4 : !moore.i32 -> !moore.i32
-    moore.blocking_assign %3, %5 : i32
-    %6 = moore.struct_extract_ref %ii, "a" : <struct<{a: i32, b: i32}>> -> <i32>
-    %7 = moore.read %x : <i32>
-    %8 = moore.constant 1 : i32
-    %9 = moore.add %7, %8 : i32
-    moore.blocking_assign %6, %9 : i32
-    %10 = moore.struct_extract %ii, "a" : <struct<{a: i32, b: i32}>> -> i32
-    moore.blocking_assign %y, %10 : i32
-    %11 = moore.struct_extract_ref %ii, "a" : <struct<{a: i32, b: i32}>> -> <i32>
-    %12 = moore.struct_extract %ii, "a" : <struct<{a: i32, b: i32}>> -> i32
-    %13 = moore.constant 1 : i32
-    %14 = moore.add %12, %13 : i32
-    moore.blocking_assign %11, %14 : i32
-    %15 = moore.struct_extract %ii, "a" : <struct<{a: i32, b: i32}>> -> i32
-    moore.blocking_assign %z, %15 : i32
-  }
-}
+// CHECK-LABEL: func.func @SplitStructs
+func.func @SplitStructs(%arg0: !moore.i42, %arg1: !moore.i1337) {
+  // Named variables
+
+  // CHECK-DAG: %x.a = moore.variable : <i42>
+  // CHECK-DAG: %x.b = moore.variable : <i1337>
+  // CHECK-NOT: moore.struct_extract_ref
+  %x = moore.variable : <struct<{a: i42, b: i1337}>>
+  %0 = moore.struct_extract_ref %x, "a" : <struct<{a: i42, b: i1337}>> -> <i42>
+  %1 = moore.struct_extract_ref %x, "b" : <struct<{a: i42, b: i1337}>> -> <i1337>
+  // CHECK: moore.blocking_assign %x.a, %arg0
+  // CHECK: moore.blocking_assign %x.b, %arg1
+  moore.blocking_assign %0, %arg0 : !moore.i42
+  moore.blocking_assign %1, %arg1 : !moore.i1337
```

Comment:
I'm curious how to handle the `moore.extract` like
```
%0 = moore.read %x;
%1 = moore.extract %0, "a"
```
when we run `--sroa`, `%x` will be earsed.

Reply:
Having a `read %x` blocks SROA for `%x`. For example, the following code without read causes SROA to expand the variable `%x`:
```mlir
func.func @Case1() {
  %x = moore.variable : <struct<{a: i42, b: i1337}>>
  moore.struct_extract_ref %x, "a" : <struct<{a: i42, b: i1337}>> -> <i42>
  moore.struct_extract_ref %x, "b" : <struct<{a: i42, b: i1337}>> -> <i1337>
  return
}
// after circt-opt --sroa
func.func @Case1() {
  %x.b = moore.variable : <i1337>
  %x.a = moore.variable : <i42>
  return
}
```
But if I add a `read %x`, SROA does not expand the variable:
```mlir
func.func @Case2() {
  %x = moore.variable : <struct<{a: i42, b: i1337}>>
  moore.struct_extract_ref %x, "a" : <struct<{a: i42, b: i1337}>> -> <i42>
  moore.struct_extract_ref %x, "b" : <struct<{a: i42, b: i1337}>> -> <i1337>
  moore.read %x : <struct<{a: i42, b: i1337}>>
  return
}
// after circt-opt --sroa
func.func @Case2() {
  %x = moore.variable : <struct<{a: i42, b: i1337}>>
  moore.struct_extract_ref %x, "a" : <struct<{a: i42, b: i1337}>> -> <i42>
  moore.struct_extract_ref %x, "b" : <struct<{a: i42, b: i1337}>> -> <i1337>
  moore.read %x : <struct<{a: i42, b: i1337}>>
  return
}
```
So it looks like SROA is doing the correct thing in case there is a `read` which it cannot properly expand.

Reply:
Yes, SROA will be enabled only if the destructurable variable is unused after rewiring.

Reply:
In this case, if we have a local variable with struct type declared in the procedure body. We cannot eliminate it using `mem2reg`. But it's not a problem, we have decided to lower `moore.procedure` to `llhd.process` :smiley: .

Reply:
Sounds good! Maybe we'll find more optimizations later on to help eliminate more of these variables -- in case that's even needed. But as long as we can lower it to LLHD, things should work :slightly_smiling_face:

Reply:
@fabianschuiki  what order should SROA and Canonicalize be?
If this PR https://github.com/llvm/circt/pull/7341 is applied, no more `variableOp` for struct will exist after the Canonicalize Pass.

Reply:
I don't think we can canonicalize variables away everywhere. There are a lot of use cases that involve multiple assignments to variables, potentially from different processes. So I'm pretty sure that canonicalization will not really remove any variables. Things like a mem2reg conversion is fairly involved and requires quite a bit of analysis. I'm pretty sure that can't be done in a canonicalization pattern alone, but requires a dedicated mem2reg pass. Does the one that MLIR already has work for this?